File: ai-orchestrator/src/main/java/com/lingxi/ai/orchestrator/service/DAGValidator.java
package com.lingxi.ai.orchestrator.service;

import com.lingxi.ai.orchestrator.model.DAGRequest;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;

import java.util.*;

@Service
public class DAGValidator {

    private static final Logger logger = LoggerFactory.getLogger(DAGValidator.class);

    public void validate(DAGRequest dag) {
        if (dag == null || dag.getNodes() == null || dag.getNodes().isEmpty()) {
            throw new IllegalArgumentException("DAG must contain at least one node");
        }

        Set<String> nodeIds = new HashSet<>();
        for (DAGRequest.NodeRequest node : dag.getNodes()) {
            if (node.getId() == null || node.getId().isEmpty()) {
                throw new IllegalArgumentException("Node id cannot be empty");
            }
            if (!nodeIds.add(node.getId())) {
                throw new IllegalArgumentException("Duplicate node id: " + node.getId());
            }
        }

        if (dag.getEdges() != null) {
            for (DAGRequest.Edge edge : dag.getEdges()) {
                if (!nodeIds.contains(edge.getFrom())) {
                    throw new IllegalArgumentException("Parent node not found: " + edge.getFrom());
                }
                if (!nodeIds.contains(edge.getTo())) {
                    throw new IllegalArgumentException("Child node not found: " + edge.getTo());
                }
            }

            detectCycle(dag, nodeIds);
            detectIsolatedNodes(dag, nodeIds);
        }

        logger.info("DAG validation passed");
    }

    private void detectCycle(DAGRequest dag, Set<String> nodeIds) {
        Map<String, List<String>> adjacencyList = new HashMap<>();
        for (String nodeId : nodeIds) {
            adjacencyList.put(nodeId, new ArrayList<>());
        }
        if (dag.getEdges() != null) {
            for (DAGRequest.Edge edge : dag.getEdges()) {
                adjacencyList.get(edge.getFrom()).add(edge.getTo());
            }
        }

        Set<String> visited = new HashSet<>();
        Set<String> recursionStack = new HashSet<>();

        for (String nodeId : nodeIds) {
            if (hasCycle(nodeId, adjacencyList, visited, recursionStack)) {
                throw new IllegalArgumentException("Cycle detected in DAG");
            }
        }
    }

    private boolean hasCycle(String nodeId, Map<String, List<String>> adjacencyList,
                              Set<String> visited, Set<String> recursionStack) {
        if (recursionStack.contains(nodeId)) {
            return true;
        }
        if (visited.contains(nodeId)) {
            return false;
        }

        visited.add(nodeId);
        recursionStack.add(nodeId);

        for (String child : adjacencyList.getOrDefault(nodeId, Collections.emptyList())) {
            if (hasCycle(child, adjacencyList, visited, recursionStack)) {
                return true;
            }
        }

        recursionStack.remove(nodeId);
        return false;
    }

    private void detectIsolatedNodes(DAGRequest dag, Set<String> nodeIds) {
        if (dag.getEdges() == null || dag.getEdges().isEmpty()) {
            if (nodeIds.size() > 1) {
                throw new IllegalArgumentException("DAG with multiple nodes must have at least one edge");
            }
            return;
        }

        Set<String> hasIncoming = new HashSet<>();
        Set<String> hasOutgoing = new HashSet<>();

        for (DAGRequest.Edge edge : dag.getEdges()) {
            hasOutgoing.add(edge.getFrom());
            hasIncoming.add(edge.getTo());
        }

        Set<String> startNodes = new HashSet<>();
        for (String nodeId : nodeIds) {
            if (!hasIncoming.contains(nodeId)) {
                startNodes.add(nodeId);
            }
        }

        if (startNodes.isEmpty()) {
            throw new IllegalArgumentException("DAG must have at least one start node (no incoming edges)");
        }

        logger.info("Found {} start nodes: {}", startNodes.size(), startNodes);
    }
}

File: internal/orchestrator/service/dag_validator.go
package service

import (
	"fmt"

	"github.com/lingxi-ai/lingxi-ai-operation-system/internal/model"
)

type DAGValidator struct{}

func NewDAGValidator() *DAGValidator {
	return &DAGValidator{}
}

func (v *DAGValidator) Validate(dag *model.DAGRequest) error {
	if dag == nil {
		return fmt.Errorf("DAG request is nil")
	}
	if len(dag.Nodes) == 0 {
		return fmt.Errorf("DAG must have at least one node")
	}

	nodeIDs := make(map[string]bool)
	for _, node := range dag.Nodes {
		if node.ID == "" {
			return fmt.Errorf("node ID cannot be empty")
		}
		if nodeIDs[node.ID] {
			return fmt.Errorf("duplicate node ID: %s", node.ID)
		}
		nodeIDs[node.ID] = true
	}

	for _, edge := range dag.Edges {
		if !nodeIDs[edge.From] {
			return fmt.Errorf("edge references unknown node: %s", edge.From)
		}
		if !nodeIDs[edge.To] {
			return fmt.Errorf("edge references unknown node: %s", edge.To)
		}
	}

	if err := v.checkCycle(dag); err != nil {
		return err
	}

	return nil
}

func (v *DAGValidator) checkCycle(dag *model.DAGRequest) error {
	adj := make(map[string][]string)
	for _, edge := range dag.Edges {
		adj[edge.From] = append(adj[edge.From], edge.To)
	}

	visited := make(map[string]int)
	var nodeIDs []string
	for _, n := range dag.Nodes {
		nodeIDs = append(nodeIDs, n.ID)
	}

	var dfs func(nodeID string) error
	dfs = func(nodeID string) error {
		if visited[nodeID] == 1 {
			return fmt.Errorf("cycle detected at node: %s", nodeID)
		}
		if visited[nodeID] == 2 {
			return nil
		}
		visited[nodeID] = 1
		for _, next := range adj[nodeID] {
			if err := dfs(next); err != nil {
				return err
			}
		}
		visited[nodeID] = 2
		return nil
	}

	for _, id := range nodeIDs {
		if visited[id] == 0 {
			if err := dfs(id); err != nil {
				return err
			}
		}
	}

	return nil
}
